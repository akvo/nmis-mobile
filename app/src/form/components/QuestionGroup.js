import React from 'react';
import { View } from 'react-native';
import { Text, Icon, Button } from '@rneui/themed';
import Question from './Question';
import { FieldGroupHeader } from '../support';
import { styles } from '../styles';
import { FormState } from '../../store';
import { i18n } from '../../lib';

const RepeatTitle = ({ index, repeat, group }) => {
  const { name } = group;

  return (
    <View style={styles.repeatTitleContainer}>
      <View style={styles.repeatTitleStart}>
        <Text testID="repeat-title">{`${name} - ${repeat + 1}`}</Text>
      </View>
      {group?.repeat > 1 && (
        <View style={styles.repeatTitleEnd}>
          <Icon
            testID="repeat-delete-button"
            name="delete"
            type="material"
            style={styles.repeatDeleteButton}
            onPress={() => console.log(index)}
          />
        </View>
      )}
    </View>
  );
};

const RepeatAddMoreButton = ({ index, group, trans }) => {
  return (
    <View style={styles.repeatAddMoreButtonContainer}>
      <Button
        style={styles.repeatAddMoreButton}
        type="outline"
        onPress={() => console.log('Add Repeat')}
        testID="repeat-add-more-button"
      >
        <Icon name="plus" type="octicon" color={'#3389DC'} style={styles.repeatAddMoreIcon} />
        {trans.addAnother}
      </Button>
    </View>
  );
};

const QuestionGroup = ({ index, group, setFieldValue, values }) => {
  const activeLang = FormState.useState((s) => s.lang);
  const trans = i18n.text(activeLang);

  const repeatable = group?.repeatable || false;
  const repeats = group?.repeats || [0];
  // const repeatText = group?.repeatText || `Number of ${name}`;

  return (
    <View>
      <FieldGroupHeader index={index} {...group} />
      {repeats.map((r) => (
        <View key={r}>
          {repeatable && <RepeatTitle index={index} repeat={r} group={group} />}
          <Question group={group} setFieldValue={setFieldValue} values={values} />
        </View>
      ))}
      {repeatable && <RepeatAddMoreButton index={index} group={group} trans={trans} />}
    </View>
  );
};

export default QuestionGroup;
