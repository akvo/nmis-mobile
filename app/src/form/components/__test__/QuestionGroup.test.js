import React from 'react';
import renderer from 'react-test-renderer';
import { render, waitFor } from '@testing-library/react-native';
import QuestionGroup from '../QuestionGroup';
import { useField } from 'formik';

jest.mock('formik', () => ({
  useField: jest.fn(),
}));

// Mock the return values of useField
const mockField = { name: 'myField', value: '', onChange: jest.fn() };
const mockMeta = { touched: false, error: '' };
const mockHelpers = { setValue: jest.fn(), setTouched: jest.fn() };

// Mock the useField function with the correct return structure
useField.mockReturnValue([mockField, mockMeta, mockHelpers]);

const question = {
  id: 1,
  name: 'Your Name',
  order: 1,
  type: 'input',
  required: true,
  meta: true,
  translations: [
    {
      name: 'Nama Anda',
      language: 'id',
    },
  ],
};

const questionGroup = {
  id: 1,
  name: 'Group 1',
  description: 'Lorem ipsum',
  order: 1,
  translations: [
    {
      name: 'Grup 1',
      language: 'id',
    },
  ],
  question: [question],
};

describe('QuestionGroup', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  test('renders correctly', async () => {
    const tree = renderer
      .create(
        <QuestionGroup
          index={1}
          group={questionGroup}
          setFieldValue={() => jest.fn()}
          values={null}
        />,
      )
      .toJSON();
    await waitFor(() => expect(tree).toMatchSnapshot());
  });

  it('should render question group name', async () => {
    const wrapper = render(
      <QuestionGroup
        index={1}
        group={questionGroup}
        setFieldValue={() => jest.fn()}
        values={null}
      />,
    );
    await waitFor(() => {
      const title = wrapper.queryByTestId('text-name');
      expect(title).toBeTruthy();
    });
  });

  it('should not render question group description if not defined', async () => {
    const questionGroupTmp = {
      ...questionGroup,
      description: null,
    };
    const wrapper = render(
      <QuestionGroup
        index={1}
        group={questionGroupTmp}
        setFieldValue={() => jest.fn()}
        values={null}
      />,
    );
    await waitFor(() => {
      const description = wrapper.queryByTestId('text-description');
      expect(description).toBeFalsy();
      expect(description).toEqual(null);
    });
  });

  it('should render question group description if any', async () => {
    const wrapper = render(
      <QuestionGroup
        index={1}
        group={questionGroup}
        setFieldValue={() => jest.fn()}
        values={null}
      />,
    );
    await waitFor(() => {
      const description = wrapper.queryByTestId('text-description');
      expect(description).toBeTruthy();
    });
  });

  it('should not render repeatable title and buttons if question group not repeatable', async () => {
    const wrapper = render(
      <QuestionGroup
        index={0}
        group={questionGroup}
        setFieldValue={() => jest.fn()}
        values={null}
      />,
    );

    const repeatTitle = wrapper.queryByTestId('repeat-title');
    expect(repeatTitle).toBeFalsy();
    const repeatDelete = wrapper.queryByTestId('repeat-delete-button');
    expect(repeatDelete).toBeFalsy();
    const repeatAddMore = wrapper.queryByTestId('repeat-add-more-button');
    expect(repeatAddMore).toBeFalsy();
  });

  it('should render repeatable title and buttons if question group repeatable', async () => {
    const questionGroupTmp = {
      ...questionGroup,
      repeatable: true,
    };
    const wrapper = render(
      <QuestionGroup
        index={0}
        group={questionGroupTmp}
        setFieldValue={() => jest.fn()}
        values={null}
      />,
    );

    const repeatTitle = wrapper.queryByTestId('repeat-title');
    expect(repeatTitle).toBeTruthy();
    const repeatDelete = wrapper.queryByTestId('repeat-delete-button');
    expect(repeatDelete).toBeFalsy();
    const repeatAddMore = wrapper.queryByTestId('repeat-add-more-button');
    expect(repeatAddMore).toBeTruthy();
  });

  it.todo('should render more question group if add more button pressed');
  it.todo('should remove a repeatable question group if delete button pressed');
});
