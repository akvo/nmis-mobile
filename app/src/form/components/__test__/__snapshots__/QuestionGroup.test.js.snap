// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuestionGroup renders correctly 1`] = `
<View>
  <View>
    <View
      style={
        {
          "borderBottomWidth": 0.5,
          "borderColor": "grey",
          "borderTopWidth": 0.5,
          "paddingHorizontal": 20,
          "paddingVertical": 15,
        }
      }
    >
      <Text
        accessibilityRole="text"
        style={
          {
            "color": "#242424",
            "fontSize": 14,
            "fontWeight": "bold",
          }
        }
        testID="text-name"
      >
        2. Group 1
      </Text>
    </View>
    <View
      style={
        {
          "backgroundColor": "#e5e7eb",
        }
      }
    >
      <Text
        accessibilityRole="text"
        style={
          {
            "color": "#242424",
            "fontSize": 14,
            "paddingHorizontal": 22,
            "paddingVertical": 8,
          }
        }
        testID="text-description"
      >
        Lorem ipsum
      </Text>
    </View>
  </View>
  <View>
    <View
      style={
        {
          "paddingHorizontal": 10,
          "paddingVertical": 10,
        }
      }
    >
      <View>
        <View>
          <View
            style={
              {
                "flex": 1,
                "flexDirection": "row",
              }
            }
          >
            <View
              style={
                {
                  "display": "flex",
                  "flexDirection": "row",
                  "fontSize": 14,
                  "fontWeight": 600,
                  "gap": 4,
                  "marginBottom": 8,
                  "paddingHorizontal": 10,
                }
              }
            >
              <Text
                accessibilityRole="text"
                style={
                  {
                    "color": "#242424",
                  }
                }
                testID="field-label"
              >
                1. Your Name
              </Text>
              <View
                collapsable={false}
              >
                <View
                  accessibilityState={
                    {
                      "busy": undefined,
                      "checked": undefined,
                      "disabled": undefined,
                      "expanded": undefined,
                      "selected": undefined,
                    }
                  }
                  accessibilityValue={
                    {
                      "max": undefined,
                      "min": undefined,
                      "now": undefined,
                      "text": undefined,
                    }
                  }
                  accessible={true}
                  collapsable={false}
                  focusable={true}
                  onBlur={[Function]}
                  onClick={[Function]}
                  onFocus={[Function]}
                  onResponderGrant={[Function]}
                  onResponderMove={[Function]}
                  onResponderRelease={[Function]}
                  onResponderTerminate={[Function]}
                  onResponderTerminationRequest={[Function]}
                  onStartShouldSetResponder={[Function]}
                />
                <Modal
                  animationType="fade"
                  hardwareAccelerated={false}
                  onShow={[Function]}
                  transparent={true}
                  visible={false}
                />
              </View>
            </View>
          </View>
          <View
            style={
              {
                "paddingHorizontal": 10,
                "width": "100%",
              }
            }
            testID="RNE__Input__view-wrapper"
          >
            <View
              collapsable={false}
              style={
                {
                  "alignItems": "center",
                  "borderBottomWidth": 0.5,
                  "borderColor": "grey",
                  "borderRadius": 5,
                  "borderWidth": 0.5,
                  "flexDirection": "row",
                  "paddingHorizontal": 10,
                  "transform": [
                    {
                      "translateX": 0,
                    },
                  ],
                }
              }
            >
              <TextInput
                editable={true}
                onChangeText={[Function]}
                placeholderTextColor="#86939e"
                style={
                  {
                    "color": "#242424",
                    "flex": 1,
                    "fontSize": 18,
                    "minHeight": 40,
                  }
                }
                testID="type-input"
                underlineColorAndroid="transparent"
              />
            </View>
            <Text
              style={
                {
                  "color": "#ff190c",
                  "fontSize": 12,
                  "margin": 5,
                }
              }
            />
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;
